src/types.ts

```diff
@@ -119,10 +119,65 @@ export type ICalSource = {
 	url: string;
 } & CalendarSourceCommon;
 
+
+/**
+ * Auth types. Currently only support Basic, but will probably support OAuth in the future.
+ */
+type BasicAuth = {
+	username: string;
+	password: string;
+} & CalendarSourceCommon;
+
+type AuthType = BasicAuth;
+
+/**
+ * Read/write mirror of a remote CalDAV backed calendar at the given URL.
+ */
+export type CalDAVSource = {
+    type: "caldav";
+    name: string;
+    url: string;
+    homeUrl: string;
+} & CalendarSourceCommon & AuthType;
+
+/**
+ * An read/write mirror of an iCloud backed calendar.
+ */
+export type ICloudSource = Omit<CalDAVSource, "type" | "url"> & {
+	type: "icloud";
+	url: "https://caldav.icloud.com";
+};
+
 export type CalendarSource =
 	| LocalCalendarSource
 	| GoogleCalendarSource
-	| ICalSource;
+	| ICalSource
+	| CalDAVSource
+	| ICloudSource;
+
+/**
+ * Construct a partial calendar source of the specified type
+ */
+export function partialCalendarSource(
```

Comment:
```suggestion
export function makeDefaultPartialCalendarSource(
```

Maybe this name is too long but feel like it'll help just to be clear that this is adding _default_ values.